### PR TITLE
Preg match regex mutators #2

### PR DESCRIPTION
**Pull request by maks-rafalko**
This PR is a rebirth of https://github.com/infection/infection/pull/388

Fixes https://github.com/infection/infection/issues/324

- took @adeptofvoltron's work
- fixed all comments and adopted to the current code base
- added more tests
- [x] Doc PR: https://github.com/infection/site/pull/203

2 new mutations here:

```diff
# 1
- preg_match("/^test$/", $string));
+ preg_match("/^test/", $string));

# 2
- preg_match("/^test$/", $string));
+ preg_match("/test$/", $string));
```

Other possible mutations will be done in a separate PRs after this one is merged.

### CODE REVIEW

File: psalm-baseline.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<files psalm-version="4.1.1@16bfbd9224698bd738c665f33039fade2a1a3977">
+  <file src="src/Mutator/Regex/PregMatchRemoveCaret.php">
+    <ImpureFunctionCall occurrences="1">
```
Comment (maks-rafalko) [inline psalm-baseline.xml:4]: Unfortunately, `thecodingmachine/safe` package does not support Psalm's annotations yet, and functions are missing `@pure` annotation, that's why we have issues reported by Psalm.

Tried https://github.com/hectorj/safe-php-psalm-plugin, but it doesn't work properly, so for now it's ok to suppress them

File: src/Mutator/Regex/AbstractPregMatch.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+        yield new FuncCall($node->name, [$newArgument] + $arguments, $node->getAttributes());
+    }
+
+    public function canMutate(Node $node): bool
+    {
+        return $node instanceof FuncCall
+            && $node->name instanceof Node\Name
+            && strtolower((string) $node->name) === 'preg_match'
+            && $node->args[0]->value instanceof Node\Scalar\String_
```
Comment (BackEndTea) [inline src/Mutator/Regex/AbstractPregMatch.php:75]: How will this handle an unpacked array as an argument? Kind of an edge case, but may be worth making sure it doesn't crash on that.,

e.g.
```php
\preg_match(...func_get_args());
```

Comment (maks-rafalko) [reply src/Mutator/Regex/AbstractPregMatch.php:75]: added a test to cover this case

only raw strings are mutated for now, with an unpacked array this mutator will do nothing

Comment (BackEndTea) [reply src/Mutator/Regex/AbstractPregMatch.php:75]: 👍

File: src/Mutator/Regex/PregMatchRemoveCaret.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        $flags = $matches[4] ?? '';
+
+        return $delimiter . $regexBody . $delimiter . $flags;
+    }
+
+    protected function isProperRegexToMutate(string $pattern): bool
+    {
+        preg_match(self::ANALYSE_REGEX, $pattern, $matches);
+
+        return array_key_exists(2, $matches) && $matches[2] === '^';
```
Comment (BackEndTea) [inline src/Mutator/Regex/PregMatchRemoveCaret.php:91]: prehaps this could be simplified to `($matches[2] ?? null) === '^'`

**Review by BackEndTea — APPROVED** [review]
1 nitpick

**Comment by maks-rafalko** [comment]
Thank you @BackEndTea
